src/librustc/middle/liveness.rs

```diff
@@ -1483,7 +1483,14 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
                         None if !body.stmts.is_empty() =>
                             match body.stmts.first().unwrap().node {
                                 hir::StmtSemi(ref e, _) => {
-                                    self.ir.tcx.expr_ty(&**e) == t_ret
+                                    match (&self.ir.tcx.expr_ty(&**e).sty, &t_ret.sty) {
+                                        // Type inference causes unsuffixed integer literals
+                                        // to be evaluated as i32 values. We still want to
+                                        // inform the user of a semicolon in this case.
+                                        (&ty::TyInt(ast::TyI32), &ty::TyInt(..)) => true,
+                                        (&ty::TyInt(ast::TyI32), &ty::TyUint(..)) => true,
```

Comment:
Won’t… this make case like

```
fn foo() -> u16 {
    1234i32;
}
```
also print the note, even though `i32` type is clearly intended?